src/mds/Server.cc

```diff
@@ -385,6 +452,22 @@ void Server::handle_client_session(MClientSession *m)
       }
     }
 
+    it = session->info.client_metadata.find("reclaim_uuid");
+    if (it != session->info.client_metadata.end()) {
+      auto uuid = it->second;
+      session->info.client_metadata.erase(it);
+      session->info.client_metadata["uuid"] = uuid;
+      if (reclaim_session(session, uuid) < 0) {
+        mds->clog->warn() << "failed to reclaim session (uuid='" << uuid << "')";
+	mds->send_message_client(new MClientSession(CEPH_SESSION_REJECT), session);
```

Comment:
If the MDS is not `up:active` and cannot reclaim the session, then the client should still be able to proceed with manual reclaim via a clustered graceperiod (for Ganesha).

I still think this needs to be a separate two-phase message sent by the client after establishing a session.

Reply:
> If the MDS is not up:active and cannot reclaim the session, then the client should still be able to proceed with manual reclaim via a clustered graceperiod (for Ganesha).

This is a bit of a digression from @ukernel's patches, but I need to make the point. The above is not 100% true.

If the client came up and it couldn't reclaim its state, then any previous session it held has expired. Once that occurs, the MDS (or another) could have handed out state that conflicts with what that session held.

At that point, a grace period will not help. Any state that was held by that client previously is now invalid, and no NFS client of that ganesha server can be allowed to reclaim anything.

A cluster-wide grace period really only helps in the situation where we can guarantee that no conflicting state was handed out after the server(s) failed. IOW, the cluster-wide grace period needs to be in effect before a client's session expires or we can't guarantee consistency.

So as a practical matter, the cluster-wide grace period is really only for catastrophic failures where we lose ganesha + its MDSs such that the only record of state is held by the NFS clients.

The problem is how to detect that when ganesha restarts? It doesn't really have a mechanism to know what happened during its outage. It would be nice to allow a single ganesha to request a grace period if it thinks it would be able to reclaim from it (and thus do this automatically in a completely distributed way), but I have yet to figure out how to make that work.

For now, establishing a cluster-wide grace period is a manual process (via a command-line tool), for that reason. Maybe we can teach kubernetes or whatever to do this for us when starting a new deployment? I think it probably will need to be handled out of band from ganesha though.

One thing I think we probably will want to do (for safety reasons) is have the MDS clear out that ganesha instance's recovery db object when its session expires. That should be a very simple call to remove an object from a pool. Ideally we'd do that before expiring the session to ensure that there are no races. Maybe we'll need way to set a rados transaction to be called by the MDS when an MDS session is expired? That would help ensure that nothing can use that DB to do recovery after that point.

Reply:
maybe we can make mds not evict ganesha session automatically. If mds fails, the ganesha session prevents mds from leaving reconnect state. then the new ganesha client start, it sends reclaim request to MDS. mds evicts the old session and continues to recovery.

Reply:
> If the client came up and it couldn't reclaim its state, then any previous session it held has expired. Once that occurs, the MDS (or another) could have handed out state that conflicts with what that session held.

Recall from the diagram in the recovery document that there are two reasons it cannot reclaim state. Using the current errors described in the doc: if the session is stale (-ESTALE, the MDS has lost the state due to recovery) or the session is invalid (-EINVAL, the rank has evicted the session at some point between now and when the session went stale).

If the session is stale, then the client must do its own recovery through a clustered graceperiod. This is the scenario I'm referring to. The mount should still succeed and the client should be able to recover.